src/pieces/messageCount.ts

```diff
@@ -1,23 +1,79 @@
-import { Client, TextChannel } from 'discord.js';
+import { Client, TextChannel, Role, Message, MessageEmbed } from 'discord.js';
 import { generateLogEmbed } from '@lib/utils';
 import { DatabaseError } from '@lib/types/errors';
-import { LOG, PREFIX, DB } from '@root/config';
+import { LOG, PREFIX, DB, ROLES } from '@root/config';
+import { SageUser } from '@lib/types/SageUser';
+
+const xpRatio = 1.25;
+const startingColor = 80;
+const greenIncrement = 8;
+const maxGreen = '00ff00';
+const maxLevel = 20;
 
 async function register(bot: Client): Promise<void> {
 	const errLog = await bot.channels.fetch(LOG.ERROR) as TextChannel;
-	bot.on('message', msg => {
+	bot.on('message', async msg => {
 		if (msg.channel.type !== 'text' || msg.content.toLowerCase().startsWith(PREFIX) || msg.author.bot) {
 			return;
 		}
 
 		const channel = msg.channel as TextChannel;
-		if (channel.topic && channel.topic.startsWith('[no message count]')) {
-			return;
+
+		const entry: SageUser = await bot.mongo.collection(DB.USERS).findOne({ discordId: msg.author.id });
+
+		if (!entry) {
+			throw `member ${msg.author.username} (${msg.author.id}) not in database`;
+		}
+
+		if (!channel.topic || (channel.topic && !channel.topic.startsWith('[no message count]'))) {
+			entry.count++;
+		}
+
+		if (--entry.curExp <= 0) {
+			entry.curExp = entry.levelExp = Math.floor(entry.levelExp * xpRatio);
+			entry.level++;
+			if (entry.levelPings) {
+				sendLevelPing(msg, entry);
+			}
+			let addRole: Role;
+			if (!(addRole = msg.guild.roles.cache.find(r => r.name === `Level ${entry.level}`))
+			&& entry.level <= maxLevel) { // make a new level role if it doesn't exist
+				addRole = await msg.guild.roles.create({
+					data: {
+						name: `Level ${entry.level}`,
+						color: createLevelHex(entry.level),
+						position: msg.guild.roles.cache.get(ROLES.VERIFIED).position + 1,
+						permissions: 0
+					},
+					reason: `${msg.author.username} is the first to get to Level ${entry.level}`
+				});
+			}
+
+			if (entry.level <= maxLevel) {
+				msg.member.roles.remove(msg.member.roles.cache.find(r => r.name.startsWith('Level')), `${msg.author.username} leveled up.`);
```

Comment:
await this so you dont somehow end up adding the new one first and then removing it if the api lags.